### PR TITLE
[Access] Expose GetAccountBalance and GetAccountKeys in script execution module

### PR DESCRIPTION
**Pull request by AndriiDiachuk**
Closes: #5894 

In this PR `ScriptExecutor` has new methods `GetAccountBalance` to get an account balance and `GetAccountKeys` keys by height.

### CODE REVIEW

File: fvm/environment/account_info.go [hunk 1]
```diff
@@ -254,3 +269,21 @@ func (info *accountInfo) GetAccount(
...
+	address flow.Address,
+) (
+	[]flow.AccountPublicKey,
+	error,
+) {
+	defer info.tracer.StartChildSpan(trace.FVMEnvGetAccount).End()
+
+	//accountKeys, err := info.accounts.GetPublicKey(address, 0)
+
+	account, err := info.accounts.Get(address)
```
Comment (peterargue) [inline fvm/environment/account_info.go:283]: maybe we can just add this to the `Accounts` interface:
https://github.com/onflow/flow-go/blob/091b7412aa23629ecca51255958d4acb94d04183/fvm/environment/accounts.go#L251

File: fvm/fvm.go [hunk 2]
```diff
@@ -242,3 +247,74 @@ func (vm *VirtualMachine) GetAccount(
 	}
 	return account, nil
 }
+
+// GetAccountBalance returns an account balance by address or an error if none exists.
+func (vm *VirtualMachine) GetAccountBalance(
```
Comment (peterargue) [inline fvm/fvm.go:252]: it doesn't look like `vm` is actually used in any of these, so we probably don't need to add them to the `VirtualMachine` interface. Instead, just make them standalong methods.

File: fvm/fvm.go [hunk 3]
```diff
@@ -242,3 +247,74 @@ func (vm *VirtualMachine) GetAccount(
...
+	env := environment.NewScriptEnv(
+		context.Background(),
+		ctx.TracerSpan,
+		ctx.EnvironmentParams,
+		storageTxn)
+
+	accountKeys, err := env.GetAccountKeys(address)
+
+	if err != nil {
+		return nil, fmt.Errorf("cannot get account balance: %w", err)
```
Comment (peterargue) [inline fvm/fvm.go:317]:
```suggestion
		return nil, fmt.Errorf("cannot get account keys: %w", err)
```

File: fvm/fvm.go [hunk 4]
```diff
@@ -242,3 +247,74 @@ func (vm *VirtualMachine) GetAccount(
...
+		storageTxn)
+	accountBalance, err := env.GetAccountBalance(common.MustBytesToAddress(address.Bytes()))
+
+	if err != nil {
+		return 0, fmt.Errorf("cannot get account balance: %w", err)
+	}
+	return accountBalance, nil
+}
+
+// GetAccountKeys returns an account balance by address or an error if none exists.
```
Comment (peterargue) [inline fvm/fvm.go:286]:
```suggestion
// GetAccountKeys returns an account keys by address or an error if none exists.
```

File: fvm/environment/account_info.go [hunk 5]
```diff
@@ -254,3 +268,20 @@ func (info *accountInfo) GetAccount(
...
 	return account, nil
 }
+
+func (info *accountInfo) GetAccountKeys(
+	address flow.Address,
+) (
+	[]flow.AccountPublicKey,
+	error,
+) {
+	defer info.tracer.StartChildSpan(trace.FVMEnvGetAccount).End()
```
Comment (peterargue) [inline fvm/environment/account_info.go:278]: can you add a new trace constant `FVMEnvGetAccountKeys` here
https://github.com/onflow/flow-go/blob/dfd15534a32ce07085fc41759bbb256d3fd1abce/module/trace/constants.go#L164

File: fvm/environment/account_info.go [hunk 6]
```diff
@@ -103,6 +104,19 @@ func (info ParseRestrictedAccountInfo) GetAccount(
...
 
+func (info ParseRestrictedAccountInfo) GetAccountKeys(
+	address flow.Address,
+) (
+	[]flow.AccountPublicKey,
+	error,
+) {
+	return parseRestrict1Arg1Ret(
+		info.txnState,
+		trace.FVMEnvGetAccount,
```
Comment (peterargue) [inline fvm/environment/account_info.go:115]: and then use the new const here as well

File: module/execution/scripts_test.go [hunk 7]
```diff
@@ -272,3 +393,52 @@ func newBlockHeadersStorage(blocks []*flow.Block) storage.Headers {
 
 	return synctest.MockBlockHeaderStorage(synctest.WithByHeight(blocksByHeight))
 }
+
+func transferTokensTx(chain flow.Chain) *flow.TransactionBody {
+	sc := systemcontracts.SystemContractsForChain(chain.ChainID())
+	return flow.NewTransactionBody().
+		SetScript([]byte(fmt.Sprintf(
```
Comment (peterargue) [inline module/execution/scripts_test.go:400]: can you reduce the indentation a bit?

**Review by peterargue — COMMENTED** [review]
Looks good. added a few small comments, but I think this is ready

**Comment by peterargue** [comment]
@janezpodhostnik can you take a look at this PR?

File: fvm/environment/accounts.go [hunk 8]
```diff
@@ -27,6 +27,7 @@ type Accounts interface {
 	AppendPublicKey(address flow.Address, key flow.AccountPublicKey) error
 	GetPublicKey(address flow.Address, keyIndex uint64) (flow.AccountPublicKey, error)
 	SetPublicKey(address flow.Address, keyIndex uint64, publicKey flow.AccountPublicKey) ([]byte, error)
+	GetPublicKeys(address flow.Address) ([]flow.AccountPublicKey, error)
 	GetContractNames(address flow.Address) ([]string, error)
```
Comment (janezpodhostnik) [inline fvm/environment/accounts.go:31]: can you change the function `func (a *StatefulAccounts) GetPublicKeys(` in this file?


it wraps a returned error:
```
	count, err := a.GetPublicKeyCount(address)
	if err != nil {
		return nil, fmt.Errorf(
			"failed to get public key count of account: %w",
			err)
	}
```
 but that error is already wrapped with that exact text

File: fvm/fvm.go [hunk 9]
```diff
@@ -242,3 +243,74 @@ func (vm *VirtualMachine) GetAccount(
 	}
 	return account, nil
 }
+
+// GetAccountBalance returns an account balance by address or an error if none exists.
+func GetAccountBalance(
```
Comment (janezpodhostnik) [inline fvm/fvm.go:248]: Can you extract the common parts of GetAccountKeys, GetAccountBalance, GetAccount there is a lot of repeated boilerplate.

File: fvm/fvm.go [hunk 10]
```diff
@@ -242,3 +233,70 @@ func (vm *VirtualMachine) GetAccount(
...
+	[]flow.AccountPublicKey,
+	error,
+) {
+	storageTxn := initializeStorageTransaction(ctx, storageSnapshot)
+
+	env := environment.NewScriptEnv(
+		context.Background(),
+		ctx.TracerSpan,
+		ctx.EnvironmentParams,
+		storageTxn)
```
Comment (janezpodhostnik) [inline fvm/fvm.go:277]: This part can be in the common function as well. The `initializeStorageTransaction` could then be renamed to something like `getScriptEnvironment`

Comment (janezpodhostnik) [reply fvm/fvm.go:277]: Sorry, I missed that.

The return type should be `environment.Environment`. The new functions can be added to `environment.Environment`.

**Review by peterargue — APPROVED** [review]

**Comment by peterargue** [comment]
@janezpodhostnik can you take another look?

**Review by AlexHentschel — APPROVED** [review]
Looks good. I have limited context on the FVM changes, but it also doesn't look like serious revisions.

**Review by janezpodhostnik — APPROVED** [review]
Re-checked. This looks good. Thank you!

**Pull request merged**